### DISCUSSION TITLE
Local development DockerFile

### DISCUSSION THREAD

**Discussion by roje-bodc** [ERDDAP/erddap#136] [Show and tell]
Hi, as part of some recent sprint work for BODC, we have created a development DockerFile that can be used to easily build and run an instance of ERDDAP locally.  This was done to remove the effort of having to manually install maven and tomcat in order for developers to get the app running and check any functional changes to ERDDAP.  This is currently sitting on our ERDDAP fork repository:

https://github.com/British-Oceanographic-Data-Centre/erddap/blob/master/development/Dockerfile

I thought it may be worth seeing if this was something you would be interested in adding to the main ERDDAP GitHub project? 
I can create a pull request if you would like to see all the relevant changes.

(Note that it is purely for local development purposes as a simple way to verify new functionality.)

Cheers,
Rob

**Comment by rmendels** [comment]
@roje-bodc  I will leave it up to Chris,  but regardless thanks for doing this.  This is great.

**Comment by ChrisJohnNOAA** [comment]
> Hi, as part of some recent sprint work for BODC, we have created a development DockerFile that can be used to easily build and run an instance of ERDDAP locally. This was done to remove the effort of having to manually install maven and tomcat in order for developers to get the app running and check any functional changes to ERDDAP. This is currently sitting on our ERDDAP fork repository:
> 
> https://github.com/British-Oceanographic-Data-Centre/erddap/blob/master/development/Dockerfile
> 
> I thought it may be worth seeing if this was something you would be interested in adding to the main ERDDAP GitHub project? I can create a pull request if you would like to see all the relevant changes.
> 
> (Note that it is purely for local development purposes as a simple way to verify new functionality.)
> 
> Cheers, Rob

Thanks for sharing. I do think that'd be good to include. Would you be able to write up instructions for using this as part of the pull request?

**Reply by roje-bodc** [reply]
Hi Chris, yes I can do that. Where would you like these instructions to be included? I could create a separate README within the directory on the DockerFile, or I could update the main README with a small section on how to use the DockerFile for development purposes? 

Else I can just attach the instructions to the pull request and you can add them where you'd like them to go.

**Reply by ChrisPJohn** [reply]
How about a separate readme in the directory with the docker file.

**Comment by srstsavage** [comment]
This is great! Very nice to have an explicit, isolated build recipe including dependencies spelled out in a Dockerfile. Two thoughts:

- Might be worth adding a [cache mount](https://docs.docker.com/build/guide/mounts/) for the m2 repository to the `mvn` commands, or at least the `mvn dependency:go-offline` invocation (e.g. `RUN --mount=type=cache,id=m2_repo,target=/root/.m2/repository mvn dependency:go-offline`). Currently the build is set up to copy the pom and download dependencies as early as possible to enable layer caching, but any change to the pom is going to result in a fresh download of every dependencies (aka "download the internet") which can take quite a while.
- If added to the main erddap repo, the development Dockerfile should be named in a way that makes it clear that it's meant for development and not deployments (`development/Dockerfile` as it is in the BODC repo, or `Dockerfile.dev`, etc).

**Reply by roje-bodc** [reply]
Hi, yes the cache mount is a good idea! I've made some alterations to the BODC DockerFile based on your suggestion. Now both the m2 repo and the download_cache directory are cached via a cache mount.

I've also updated the way the netCDF jar is added. There is an Ant run job in the pom.xml to download this, but that doesn't seem to work correctly for me. Hence why I am manually downloading the netCDF jar and adding it to the lib folder in order to get it to skip the download part of that job.

https://github.com/British-Oceanographic-Data-Centre/erddap/commit/e71c68639da7a4d2bf5058dd5640b62fe5d0b977

**Comment by roje-bodc** [comment]
@ChrisJohnNOAA  I have created a pull request for the changes here: https://github.com/ERDDAP/erddap/pull/137

Note there were some small unit test changes made that I have outlined in the PR.

Rob

**Reply by ChrisJohnNOAA** [reply]
Thanks! It's been merged in.

**Reply by roje-bodc** [reply]
Cheers Chris, I'll close this discussion now.

**Comment by roje-bodc** [comment]
These changes have now been merged into the main ERDDAP branch https://github.com/ERDDAP/erddap/pull/137 .
